pseudopotential_calculator/calculations/slab.py

```diff
@@ -26,36 +23,100 @@
+    heights = _get_sorted_heights(atom)
+    return heights[-2]
+
+
+def get_n_slab_layer(atoms: Atoms) -> int:
+    heights = _get_sorted_heights(atoms)
+    return len(heights)
+
+
+def get_slab_width(atoms: Atoms) -> int:
+    # n_slab_atoms / n_atoms_per_repeat gives the number of slab units set to be repeated.
+    # take square root to get width since the repeat s set to be the same in x and y directions
+    n_slab_atoms = len(atoms.positions) - 1  # type: ignore bad lib
+    n_atoms_per_repeat = get_n_slab_layer(atoms) - 1
+    return np.sqrt(round(n_slab_atoms / n_atoms_per_repeat))
+
+
+def _get_n_vacuum_layers_from_slab(atoms: Atoms) -> float:
+    cell_height = atoms.get_cell().array[-1][-1]
+
+    slab_height = get_height_of_layer(atoms, 1)
+    height_per_layer = get_average_height_of_layer(atoms)
+
+    vacuum_height = cell_height - slab_height
+
+    return vacuum_height / height_per_layer
+
+
+def _get_fixed_atom_indices(atoms: Atoms) -> list[int]:
+    fixed_indices = list[int]()
+    for constraint in atoms.constraints:  # type: ignore bad lib
+        if isinstance(constraint, FixAtoms):
+            fixed_indices.append(constraint.index[0])  # type: ignore bad lib
+    return fixed_indices
 
-def get_slab_calculator(
-    atoms: Atoms,
-    parameters: SlabOptimizationParams,
-    config: CastepConfig,
-    *,
-    task: Literal["Energy", "GeometryOptimization"] = "Energy",
-) -> Castep:
-    calculator = get_default_calculator(config)
 
-    calculator.param.xc_functional = parameters.xc_functional
-    calculator.param.cut_off_energy = parameters.cut_off_energy
-    calculator.param.spin_polarized = parameters.spin_polarized
-    calculator.cell.kpoint_mp_grid = parameters.kpoint_mp_grid
+def _get_heights(atoms: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
+    return atoms.positions[:, 2]  # type: ignore bad lib
 
-    # Prevent the bulk cell from rotating
-    calculator.cell.cell_constraints = "0 0 0\n0 0 0"
-    calculator.task = task
 
-    calculator.set_atoms(atoms)  # type: ignore unknown
-    # Temporary fix for bug in ase
-    atoms.calc = calculator
-    return calculator
+def _get_layer_heights(atoms: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
+    heights = _get_heights(atoms)
+    return np.unique(heights)
+
+
+def _get_layer_indices(atoms: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
+    heights = _get_layer_heights(atoms)
+    return cast(np.ndarray[Any, np.dtype[np.float64]], np.argsort(heights))
+
+
+def get_height_per_fixed_layer(atoms: Atoms) -> float:
+    heights = _get_layer_heights(atoms)
+    fixed_index = _get_fixed_atom_indices(atoms)
+    atom_0 = fixed_index[0]
+    atom_1 = fixed_index[1]
+    layer_indices = _get_layer_indices(atoms)
+    return (heights[atom_0] - heights[atom_1]) / (
+        layer_indices[atom_0] - layer_indices[atom_1]
+    )
+
+
+def get_average_height_of_layer(atoms: Atoms) -> float:
+    heights = _get_layer_heights(atoms)
```

Comment:
use get_height_per_fixed_layer in vacuum calcualtion. make per_fixed_layer default to fixed layers but fall back to any layer ( and rename to get_average_layer_height)